### PR TITLE
Latex Rendering

### PR DESCRIPTION
**Pull request by prairir**
A full way to render and then get an image of latex in MathMode
using the command **~latex** anything after that command will get put into a file, compiled to .dvi, then converted to an image, and sent back. 

with this change people can fully render math questions or anything to do with math.

the Requirments are latex and dvipng

### CODE REVIEW

File: secrets-example.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
   "image_directory_url": "http://google.ca/images",
   "api_token": "token",
-  "api_client_id": "client",
+  "api_client_id": "client"
```
Comment (EricPickup) [inline secrets-example.json:4]: I noticed this earlier today, thanks for fixing it

File: services/latex_service.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      return false
+    end
+
+    write2file(message, path, file)
+
+    # latex in nonstopmode so if error it just quits
+    # put output into /dev/null gets rid of output
+    did_comp = system("latex -interaction=nonstopmode -output-directory=#{path} #{File.join(path, file)}.tex >>/dev/null")
+
+
```
Comment (EricPickup) [inline services/latex_service.rb]: No need for two spaces, one is fine
```suggestion
```

File: services/latex_service.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    end
+  end
+
+  # writes the message to the file
+  def self.write2file(message, path, file)
+    # reading the template file
+    template = File.read(File.join(path, 'template.tex'))
+
+    # changing the file template and writing it to a new file
+    File.open(File.join(path, file + '.tex'), 'w'){ |outfile|
```
Comment (EricPickup) [inline services/latex_service.rb]:
```suggestion
    File.open(File.join(path, file + '.tex'), 'w') { |outfile|
```

File: services/latex_service.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+  # deletes the extra files
+  def self.cleanup(path, file)
+    # these are the files that are created
+    file_endings = ['.aux', '.log', '.dvi', '.png', '.tex']
+    file_endings.each{ |fending|
+      if File.exist?(File.join(path, file + fending))
+        File.delete(File.join(path, file + fending))
+      end
+    }
```
Comment (EricPickup) [inline services/latex_service.rb]: Try this syntax: 
```suggestion
    file_endings.each do |fending|
        File.delete(File.join(path, file + fending)) if File.exist? (File.join(path, file + fending))
    end
```

File: services/latex_service.rb [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  # sanitizes the message by putting a backslash in front of some chars
+  def self.sanitize(message)
+    # each symbol is a different one that could cause problems
+    # $ is to enter/exit math mode which would cause compilation problems
+    # \\ is obvious
+    # " could cause an escape of the latex function
+    res_chars = ['$', '\\', '"']
+    res_chars.each{ |res|
+      message = message.gsub(res, "\\#{res}")
+    }
```
Comment (EricPickup) [inline services/latex_service.rb]: Similar syntax change with these loops
```suggestion
    res_chars.each do |res|
      message = message.gsub(res, "\\#{res}")
    end
```

File: services/latex_service.rb [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    message
+  end
+
```
Comment (EricPickup) [inline services/latex_service.rb]:
```suggestion
```

File: services/latex_service.rb [hunk 7]
```diff
@@ -0,0 +1,66 @@
+class LatexService
+
+  # renders the message
+  def self.render?(message, path, file)
+    # stripping it so you cant just put in one letter or a string of backslashs
+    if message.strip.length == 1 || message.strip == '\\' * message.length
+      return false
+    end
```
Comment (EricPickup) [inline services/latex_service.rb]:
```suggestion
   return false if message.strip.length == 1 || message.strip == '\\' * message.length
```

File: main.rb [hunk 8]
```diff
@@ -45,6 +48,29 @@ class Main
...
+      # else return error
+      if LatexService.render?(clean_args, LATEX_DIRECTORY_RELATIVE_PATH, 'formula')
+        event.send_file(File.open(File.join(LATEX_DIRECTORY_RELATIVE_PATH, 'formula.png'), 'r'))
+      else
+        return_error(event.channel, 'Formula Didnt Compile')
+      end
+
+      # delete the files created
+      LatexService.cleanup(LATEX_DIRECTORY_RELATIVE_PATH, 'formula')
+
```
Comment (EricPickup) [inline main.rb]:
```suggestion
```

**Review by EricPickup — CHANGES_REQUESTED** [review]
This will be an awesome addition, thanks @prairir 🎉 Working on setting it up locally atm.

I left some change requests for syntax stuff. Mostly extra spacing and the way you do your loop blocks.

File: services/latex_service.rb (hunk 4 above)
Comment (prairir) [reply services/latex_service.rb]: I tried it before, it felt too long but I guess not. Thank you.

Comment (EricPickup) [reply services/latex_service.rb]: Rule of thumb is 120 character max in one line, you can add a ruler in most IDE's :)

**Comment by prairir** [comment]
I believe I fixed all the syntax changes that you requested.

**Review by EricPickup — APPROVED** [review]
Awesome stuff @prairir 🚢

**Pull request merged**